spark/src/main/java/org/apache/iceberg/spark/source/SparkPlannerUtil.java

```diff
@@ -0,0 +1,231 @@
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.MapFunction;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.types.StructType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SparkPlannerUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(SparkPlannerUtil.class);
+  private static final Types.StructType EMPTY_STRUCT = Types.StructType.of();
+
+  public static CloseableIterable<FileScanTask> planFiles(SparkSession spark, TableScan scan) {
+    Snapshot snapshot = scan.snapshot();
+    if (snapshot != null) {
+      LOG.info("Distributed Scanning table {} snapshot {} created at {} with filter {}", scan.table(),
+          snapshot.snapshotId(), snapshot.timestampMillis(), scan.filter());
+
+      Listeners.notifyAll(
+          new ScanEvent(scan.table().toString(), snapshot.snapshotId(), scan.filter(), scan.schema()));
+
+      return planFilesFor(spark, scan, snapshot);
+    } else {
+      LOG.info("Scanning empty table {}", scan.table());
+      return CloseableIterable.empty();
+    }
+  }
+
+  private static CloseableIterable<FileScanTask> planFilesFor(SparkSession spark, TableScan scan, Snapshot snapshot) {
+    //This is different for different BaseTableScan implementations
+    if (scan instanceof DataTableScan) {
+      return planDataTableScanFiles(spark, (DataTableScan) scan, snapshot);
+    } else {
+      throw new UnsupportedOperationException(String.format("Unable to do distributed planning for %s",
+          scan.getClass()));
+    }
+  }
+
+  //TODO these two functions probably belong in util now
+  private static String metadataTableName(Table table, MetadataTableType type) {
+    return metadataTableName(table, table.toString(), type);
+  }
+
+  private static String metadataTableName(Table table, String tableName, MetadataTableType type) {
+    if (tableName.contains("/")) {
+      return tableName + "#" + type;
+    } else if (tableName.startsWith("hadoop.")) {
+      // for HadoopCatalog tables, use the table location to load the metadata table
+      // because IcebergCatalog uses HiveCatalog when the table is identified by name
+      return table.location() + "#" + type;
+    } else if (tableName.startsWith("hive.")) {
+      // HiveCatalog prepend a logical name which we need to drop for Spark 2.4
+      return tableName.replaceFirst("hive\\.", "") + "." + type;
+    } else {
+      return tableName + "." + type;
+    }
+  }
+
+  private static CloseableIterable<FileScanTask> planDataTableScanFiles(
+      SparkSession spark,
+      DataTableScan scan,
+      Snapshot snapshot) {
+
+    //TODO Currently this approach reads all manifests, no manifest filtering - Maybe through pushdowns or table read
```

Comment:
If we go this route, we need to take into account partition predicates in `rowFilter` and apply them to manifests in the scan.